### ISSUE TITLE
unused: Aliased type of a generic reports as unused when used

### ISSUE THREAD

**Issue by dlm** [golangci/golangci-lint#3492]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

When running code with generics and aliasing, I get an "unused" error depending on where the generic is aliased.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.50.0 built from (unknown, mod sum: "h1:+Xmyt8rKLauNLp2gzcxKMN8VNGqGc5Avc2ZLTwIOpEA=") on (unknown)
```

</details>

### Configuration file

<details>

```console
$ cat ./golangci.yaml
linters:
  enable:
    - unused
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.19 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/dave/.cache/go-build"
GOENV="/home/dave/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS="-buildvcs=false"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/dave/dev/go/pkg/mod"
GONOPROXY="github.com/blocky/*"
GONOSUMDB="github.com/blocky/*"
GOOS="linux"
GOPATH="/home/dave/dev/go"
GOPRIVATE="github.com/blocky/*"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/dave/tmp/dev/go/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build4184969621=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run --config ./golangci.yaml -v

INFO [config_reader] Used config file golangci.yaml 
INFO [lintersdb] Active 7 linters: [errcheck gosimple govet ineffassign staticcheck typecheck unused] 
INFO [loader] Go packages loading at mode 575 (compiled_files|deps|imports|exports_file|files|name|types_sizes) took 59.330592ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 208.689µs 
INFO [linters_context/goanalysis] analyzers took 1.176332393s with top 10 stages: buildir: 916.567926ms, inspect: 58.367022ms, ctrlflow: 38.038894ms, printf: 36.281682ms, fact_deprecated: 32.266716ms, SA5012: 28.269811ms, fact_purity: 26.955866ms, nilness: 17.192928ms, typedness: 14.521069ms, errcheck: 381.439µs 
INFO [runner] Processors filtering stat (out/in): exclude: 1/1, path_shortener: 1/1, severity-rules: 1/1, cgo: 1/1, autogenerated_exclude: 1/1, exclude-rules: 1/1, nolint: 1/1, uniq_by_line: 1/1, max_same_issues: 1/1, path_prefixer: 1/1, filename_unadjuster: 1/1, skip_dirs: 1/1, max_from_linter: 1/1, source_code: 1/1, sort_results: 1/1, skip_files: 1/1, identifier_marker: 1/1, diff: 1/1, max_per_file_from_linter: 1/1, path_prettifier: 1/1 
INFO [runner] processing took 242.311µs with stages: nolint: 76.438µs, identifier_marker: 36.492µs, autogenerated_exclude: 31.476µs, path_prettifier: 29.773µs, source_code: 21.657µs, exclude-rules: 21.643µs, skip_dirs: 8.683µs, uniq_by_line: 4.261µs, max_same_issues: 2.796µs, path_shortener: 1.717µs, cgo: 1.527µs, max_from_linter: 1.417µs, filename_unadjuster: 1.139µs, max_per_file_from_linter: 1.008µs, skip_files: 512ns, diff: 431ns, exclude: 415ns, severity-rules: 387ns, sort_results: 329ns, path_prefixer: 210ns 
INFO [runner] linters took 907.969176ms with stages: goanalysis_metalinter: 907.624958ms 
bar/bar_test.go:16:7: type `Generic2` is unused (unused)
        type Generic2 = bar.GenericInterface[int, int]
             ^
INFO File cache stats: 1 entries of total size 341B 
INFO Memory: 11 samples, avg is 156.7MB, max is 230.6MB 
INFO Execution took 973.094095ms 
```

</details>

### Code example or link to a public repository

<details>

I created a minimal example.  The following code fails with the information described above.

```go
-- bar/bar.go --
package bar

import "testing"

func Generic[T any](*testing.T, any) {}

type GenericInterface[X any, Y any] interface {
        Foo(X) Y
}

type SpecificIntString struct{}

func (*SpecificIntString) Foo(int) string { return "" }


-- bar/bar_test.go --
package bar_test

import (
        "testing"
        "bky.sh/linterr/bar"
)

func Test2(t *testing.T) {
        type Generic2 = bar.GenericInterface[int, int]
        bar.Generic[Generic2](t, new(bar.SpecificIntString))
}
```

Using the same `bar/bar.go`  but changing `bar/bar_test.go` fails:

```go
-- bar/bar_test.go --
package bar_test

import (
        "testing"
        "bky.sh/linterr/bar"
)

// added code starts here
type Generic1 = bar.GenericInterface[int, int]

func Test1(t *testing.T) {
        bar.Generic[Generic1](t, new(bar.SpecificIntString))
}
// and ends here

func Test2(t *testing.T) {
        type Generic2 = bar.GenericInterface[int, int]
        bar.Generic[Generic2](t, new(bar.SpecificIntString))
}
```

</details>

**Comment by ldez** [comment]
Hello,

> [x] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.).

`unused` is a part of the set of rules that comes from [staticcheck](https://github.com/dominikh/go-tools).

```console
$ staticcheck ./...                                                      
bar/bar_test.go:10:7: type Generic2 is unused (U1000)
```

**Comment by dominikh** [comment]
Various generics-related bugs in unused will be fixed in the next release, including this one.

**Comment by dlm** [comment]
@dominikh thank you for taking a look at it.   Since it is fixed in the next release, would you like me to close the issue?

**Comment by Antonboom** [comment]
Fixed in staticcheck [v0.4.0](https://github.com/dominikh/go-tools/releases/tag/2023.1).

Tested on [v0.4.3](https://github.com/dominikh/go-tools/releases/tag/2023.1.3) – no errors.
`golangci-lint` uses the same [version](https://github.com/golangci/golangci-lint/blob/59f1c6e455a6b288d859662a54707e523a05989d/go.mod#L114).
